quic/s2n-quic-platform/src/io/testing/model.rs

```diff
@@ -0,0 +1,303 @@
+                return 0;
+            }
+
+            let mut packet = packet.into_owned();
+
+            if corrupt_rate > 0 && super::rand::gen::<u64>() < corrupt_rate {
+                // randomly truncate the payload
+                let num_bytes = super::rand::gen_range(0..packet.payload.len());
+                if num_bytes > 0 {
+                    packet.payload.truncate(num_bytes);
+                }
+
+                // randomly swap bytes in the payload
+                let num_bytes = super::rand::gen_range(0..packet.payload.len());
+                if num_bytes > 0 {
+                    super::rand::swap_count(&mut packet.payload, num_bytes);
+                }
+
+                // randomly rewrite bytes in the payload
+                let num_bytes = super::rand::gen_range(0..packet.payload.len());
+                if num_bytes > 0 {
+                    for _ in 0..num_bytes {
+                        let index = super::rand::gen_range(0..packet.payload.len());
+                        packet.payload[index] = super::rand::gen();
+                    }
+                }
+            }
+
+            if !jitter.is_zero() {
+                // add a delay for the next packet to be transmitted
+                *transmit_time += gen_jitter(jitter);
+            }
+
+            // copy the transmit time for this packet
+            let mut transmit_time = *transmit_time;
+
+            if !network_jitter.is_zero() {
+                let jitter = gen_jitter(network_jitter);
+
+                // randomly add or subtract the network jitter
+                if super::rand::gen() {
+                    transmit_time += jitter;
+                } else {
+                    transmit_time = transmit_time.checked_sub(jitter).unwrap_or(now);
+                }
+            }
+
+            // reverse the addresses so the dst/src are correct for the receiver
+            packet.switch();
+
+            let model = self.clone();
+            model.0.current_inflight.fetch_add(1, Ordering::SeqCst);
+            let buffers = buffers.clone();
+
+            // spawn a task that will push the packet onto the receiver queue at the transit time
+            super::spawn(async move {
+                if now != transmit_time {
+                    super::time::delay_until(transmit_time).await;
+                }
+
+                buffers.rx(*packet.path.local_address, |queue| {
+                    model.0.current_inflight.fetch_sub(1, Ordering::SeqCst);
+                    queue.receive(packet);
+                });
+            });
+
+            1
+        };
+
+        let mut transmission_count = 0;
+        buffers.pending_transmissions(|packet| {
+            // drop packets that exceed the maximum number of inflight packets for the network
+            if self.inflight() >= self.max_inflight() {
+                return Ok(());
+            }
+
+            // retransmit the packet until the rate fails or we retransmit 5
```

Comment:
where does the 5 come from?

Reply:
Just an arbitrary limit to prevent the case where we keep retransmitting endlessly. Not sure if it's worth adding a knob for since we'd be handling 5 retransmissions the same as 10 (and even 5 is probably overkill)